lib/_http_server.js

```diff
@@ -227,6 +227,13 @@ function writeHead(statusCode, reason, obj) {
 
   var statusLine = 'HTTP/1.1 ' + statusCode + ' ' + this.statusMessage + CRLF;
 
+  // Any response with a 304 (Not Modified) status code is always 
+  // terminated by the first empty line after the header fields, 
+  // regardless of the header fields present in the message, and 
+  // thus cannot contain a message body or 'trailers' octets.
+  if (statusCode === 304 && headers && headers['Trailer'] !== undefined)
```

Comment:
Actually, now that I think more about it, just checking for `'Trailer'` won't be enough because the letters could be in any case. I wonder if a better solution would be to just check for `statusCode === 304 && state.trailer` inside _http_outgoing.js in `_storeHeader()`? Or perhaps it could be merged with the existing 204/304 check that's already being done in there.

I also don't know if we should throw or just discard the trailer header...
Thoughts @nodejs/http ?